packages/@expo/cli/src/start/server/metro/createExpoFallbackResolver.ts

```diff
@@ -0,0 +1,160 @@
+      });
+    }
+    dependencies.push(...peerDependencies);
+  }
+  // We deduplicate the dependencies and exclude modules that we know are only used for scripts or config-plugins
+  dependencies = dependencies.filter((moduleName, index, dependenciesArr) => {
+    if (EXCLUDE_ORIGIN_MODULES[moduleName]) return false;
+    return dependenciesArr.indexOf(moduleName) === index;
+  });
+  // Return test regex for dependencies and full origin module path to resolve through
+  const originModulePath = path.dirname(filePath);
+  return dependencies.length
+    ? { originModulePath, moduleTestRe: dependenciesToRegex(dependencies) }
+    : null;
+};
+
+/** Creates a fallback module resolver that resolves dependencis of modules named in `originModuleNames` via their path.
+ * @remarks
+ * The fallback resolver targets modules dependended on by modules named in `originModuleNames` and resolves
+ * them from the module root of these origin modules instead.
+ * It should only be used as a fallback after normal Node resolution (and other resolvers) have failed for:
+ * - the `expo` package
+ * - the `expo-router` package
+ * Dependencies mentioned as either optional peer dependencies or direct dependencies by these modules may be isolated
+ * and inaccessible via standard Node module resolution. This may happen when either transpilation adds these
+ * dependencies to other parts of the tree (e.g. `@babel/runtime`) or when a dependency fails to hoist due to either
+ * a corrupted dependency tree or when a peer dependency is fulfilled incorrectly (e.g. `expo-asset`)
+ * @privateRemarks
+ * This does NOT follow Node resolution and is *only* intended to provide a fallback for modules that we depend on
+ * ourselves and know we can resolve (via expo or expo-router)!
+ */
+export function createFallbackModuleResolver({
+  originModuleNames,
+  getStrictResolver,
+}: {
+  originModuleNames: string[];
+  getStrictResolver: StrictResolverFactory;
+}): ExpoCustomMetroResolver {
+  const _moduleDescriptionsCache: Record<string, ModuleDescription | null> = {};
+
+  const getModuleDescription = (
+    immutableContext: ResolutionContext,
+    platform: string | null,
+    originModuleName: string
+  ) => {
+    if (_moduleDescriptionsCache[originModuleName] !== undefined) {
+      return _moduleDescriptionsCache[originModuleName];
+    }
+    // Resolve the origin module itself through `<module>/.`
+    const context: ResolutionContext = {
+      ...immutableContext,
+      originModulePath: `${originModuleName}/package.json`,
+    };
+    return (_moduleDescriptionsCache[originModuleName] = getModuleDescriptionWithResolver(
+      getStrictResolver(context, platform),
+      originModuleName
+    ));
+  };
+
+  return function requestFallbackModule(immutableContext, moduleName, platform) {
+    // Early return if `moduleName` cannot be a module specifier
+    if (moduleName[0] === '.' || moduleName[0] === '/') {
```

Comment:
Are there not more rules here? Like `~`?

Reply:
Do we need to account for windows here? Also is this an application-only feature, e.g. should we skip when the origin is in a node_module https://github.com/expo/expo/blob/48b2c22ebaeac472f18f403d191ae7408895232a/packages/%40expo/cli/src/utils/tsconfig/resolveWithTsConfigPaths.ts#L24-L35

Reply:
@EvanBacon The module import specifiers in Metro are always resolved as POSIX formatted paths (with forward slashes). Metro does not support importing something like `import '.\path\to\file'`, meaning we don't have to account for that here.

Only when actually resolving, and using the paths to infer other things, we need to take Windows vs normal paths into account.